dspace-api/src/main/java/org/dspace/core/ProvenanceProvider.java

```diff
@@ -0,0 +1,283 @@
+import java.util.stream.Collectors;
+
+import org.apache.commons.collections4.CollectionUtils;
+import org.apache.commons.lang.StringUtils;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.dspace.app.bulkaccesscontrol.model.AccessCondition;
+import org.dspace.app.bulkaccesscontrol.model.BulkAccessControlInput;
+import org.dspace.authorize.AuthorizeException;
+import org.dspace.authorize.ResourcePolicy;
+import org.dspace.authorize.factory.AuthorizeServiceFactory;
+import org.dspace.authorize.service.ResourcePolicyService;
+import org.dspace.content.Bitstream;
+import org.dspace.content.Bundle;
+import org.dspace.content.Collection;
+import org.dspace.content.DSpaceObject;
+import org.dspace.content.Item;
+import org.dspace.content.MetadataField;
+import org.dspace.content.MetadataSchemaEnum;
+import org.dspace.content.MetadataValue;
+import org.dspace.content.clarin.ClarinLicenseResourceMapping;
+import org.dspace.content.factory.ClarinServiceFactory;
+import org.dspace.content.factory.ContentServiceFactory;
+import org.dspace.content.service.BitstreamService;
+import org.dspace.content.service.ItemService;
+import org.dspace.content.service.clarin.ClarinItemService;
+import org.dspace.content.service.clarin.ClarinLicenseResourceMappingService;
+
+/**
+ * ProvenanceProviderrest is responsible for creating provenance metadata for items based on the actions performed.
+ *
+ * @author Michaela Paurikova (dspace at dataquest.sk)
+ */
+public class ProvenanceProvider {
+    private static final Logger log = LogManager.getLogger(ProvenanceProvider.class);
+
+    private ItemService itemService = ContentServiceFactory.getInstance().getItemService();
+    private ResourcePolicyService resourcePolicyService =
+            AuthorizeServiceFactory.getInstance().getResourcePolicyService();
+
+    private ClarinItemService clarinItemService = ClarinServiceFactory.getInstance().getClarinItemService();
+    private ClarinLicenseResourceMappingService clarinResourceMappingService =
+            ClarinServiceFactory.getInstance().getClarinLicenseResourceMappingService();
+    private BitstreamService bitstreamService = ContentServiceFactory.getInstance().getBitstreamService();
+
+    private final ProvenanceMessageProvider messageProvider = new ProvenanceMessageProvider();
+
+    private void addProvenanceMetadata(Context context, Item item, String msg)
+            throws SQLException, AuthorizeException {
+        itemService.addMetadata(context, item, MetadataSchemaEnum.DC.getName(),
+                "description", "provenance", "en", msg);
+        itemService.update(context, item);
+    }
+
+    private String extractAccessConditions(List<AccessCondition> accessConditions) {
+        return accessConditions.stream()
+                .map(AccessCondition::getName)
+                .collect(Collectors.joining(";"));
+    }
+
+    private Item getItem(Context context, Bitstream bitstream) throws SQLException {
```

Comment:
?

Reply:
1. Renamed to `findItemByBitstream`.
2. This method call a `clarinItemService.findByBitstreamUUID(...)`  method and process the response - log if there is no Item,..